Update Longhorn Backing Image IP Plan for Storage Network (#406)
harvester/docs: docs/advanced/storagenetwork.md

Harvester uses Longhorn as its built-in storage system to provide block device volumes for VMs and Pods. If the user wishes to isolate Longhorn replication traffic from the Kubernetes cluster network (i.e. the management network) or other cluster-wide workloads. Users can allocate a dedicated storage network for Longhorn replication traffic to get better network bandwidth and performance.

> For more informations, please refer [Longhorn Storage Network](https://longhorn.io/docs/1.4.3/advanced-resources/deploy/storage-network/)

Author: Updated per suggestion, thanks.

    - `kubectl get -A vmi`
- All pods that are attached to Longhorn Volumes should be stopped.
    - Users could skip this step with the Harvester Storage Network setting. Harvester will stop Longhorn-related pods automatically.

> - All the on-going image uploadings or downloadings shoulde be completed or deleted.

Author: Updated per suggestion, thanks.

	- Please refer Networking page for more details and configure `Cluster Network` and `VLAN Config` but not `Networks`.
- IP range for Storage Network
	- IP range should not conflict or overlap with Kubernetes cluster networks(`10.42.0.0/16`, `10.43.0.0/16`, `10.52.0.0/16` and `10.53.0.0/16` are reserved).

> 	- IP range should be IPv4 CIDR format and Longhorn pods use Storage Network as follows:

Reviewer: ```suggestion
	- IP range should be in IPv4 CIDR format and Longhorn pods use Storage Network as follows:
```

> 	- IP range should be IPv4 CIDR format and Longhorn pods use Storage Network as following:

Author: Updated per suggestion, thanks.

> 	- IP range should be IPv4 CIDR format and Longhorn pods use Storage Network as following:

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308519157, thanks.

>     - `instance-manger-e` and `instance-manager-r` pods: 2 IPs per node, and the upgrade will require two versions of it (the old version ones will be tear down after upgrade success).

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308525772, thanks.

>     - `instance-manger-e` and `instance-manager-r` pods: These require 2 IPs per node. During an upgrade, two versions of these pods will exist (old and new), with the old version being deleted once the upgrade is successful.

Reviewer: ```suggestion
    - `instance-manger-e` and `instance-manager-r` pods: These require 2 IPs per node. During an upgrade, two versions of these pods will exist (old and new), and the old version will be deleted once the upgrade is successful.
```

>     - `backing-image-ds` pods: for processing every on-the-fly uploading/downloading backing image data soure. These pods would be tear down as the images complete downloading or uploading.

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308525772, thanks.

>     - `backing-image-ds` pods: These are employed to process on-the-fly uploading and downloading of backing image data sources. These pods will be removed once image uploads or downloads are completed.

Reviewer: ```suggestion
    - `backing-image-ds` pods: These are employed to process on-the-fly uploads and downloads of backing image data sources. These pods will be removed once image uploads or downloads are completed.
```

>     - `backing-image-manager` pods: 1 IP per disk, and the upgrade will require two versions of it (the old version ones will be tear down after upgrade success).

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308525772, thanks.

>     - `backing-image-manager` pods: 1 IP per disk, similar to the instance manager pods, two versions of these will coexist during an upgrade, and the old ones will be removed after the successful upgrade

Reviewer: ```suggestion
    - `backing-image-manager` pods: 1 IP per disk, similar to the instance manager pods. Two versions of these will coexist during an upgrade, and the old ones will be removed after the upgrade is completed.
```

>     - The required number of IPs could be calulated from a simple formula: `Required Number of IPs = Number of Nodes * 4 + Number of Disks * 2 + Number of Uploading/Downloading images`

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308525772, thanks.

>     - The required number of IPs can be calculated using a simple formula: `Required Number of IPs = Number of Nodes * 4 + Number of Disks * 2 + Number of Uploading/Downloading images`

Reviewer: ```suggestion
    - The required number of IPs is calculated using a simple formula: `Required Number of IPs = Number of Nodes * 4 + Number of Disks * 2 + Number of Images to Download/Upload`
```

> 	- For example, if your cluster has 5 nodes, each node has two disks and ten images are going to be uploaded simultaneously. The IP range should be larger than `/26` (`5 * 4 + 5 * 2 * 2 + 10 = 50`).

Reviewer: ```suggestion
	- For example, if your cluster has 5 nodes, each node has 2 disks and 10 images will be uploaded simultaneously, the IP range should be larger than `/26` (5 x 4 + 5 x 2 x 2 + 10 = 50).
```

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308525772, thanks.

> 	- For example, if your cluster has 5 nodes, each node has two disks and ten images are going to be uploaded simultaneously. The IP range should be larger than `/26` (`5 * 4 + 5 * 2 * 2 + 10 = 50`).

Author: Updated per suggestion, thanks.

> 	- For example, if your cluster has five nodes, each node has two disks, and ten images will be uploaded simultaneously. The IP range should be greater or equal to `/26` (`5 * 4 + 5 * 2 * 2 + 10 = 50`).

Reviewer: ```suggestion
	- For example, if your cluster has five nodes, each node has two disks, and ten images will be uploaded simultaneously, the IP range should be greater than or equal to `/26` (`5 * 4 + 5 * 2 * 2 + 10 = 50`).
```

Harvester will also create a new NetworkAttachmentDefinition and update the Longhorn Storage Network setting.

> Once Longhorn setting is updated, Longhorn will restart all `instance-manager-r`, `instance-manager-e`  and `backing-image-manager` pods to apply the new network configuration, and Harvester will restart pods.

Reviewer: ```suggestion
Once the Longhorn setting is updated, Longhorn will restart all `instance-manager-r`, `instance-manager-e`, and `backing-image-manager` pods to apply the new network configuration, and Harvester will restart the pods.
```

Author: Updated per suggestion, thanks.

#### Step 2

> Check if all Longhorn `instance-manager-e`, `instance-manager-r` and `backing-image-manager` pods are ready and networks are correct.

Reviewer 2: ```suggestion
Verify the readiness of all Longhorn `instance-manager-e`, `instance-manager-r`, and `backing-image-manager` pods, and confirm that their networks are correctly configured.

Execute the following command to inspect a pod's details:

```

Author: Updated per suggestion, thanks.

> Check if all Longhorn `instance-manager-e`, `instance-manager-r` and `backing-image-manager` pods are ready and networks are correct.

Reviewer: ```suggestion
Check if all Longhorn `instance-manager-e`, `instance-manager-r`, and `backing-image-manager` pods are ready and networks are correct.
```

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308530167, thanks.

```bash
kubectl -n longhorn-system describe pod <pod-name>
```

> If it appears similar event as following, the Storage Network should run out of IPs.

Reviewer 2: ```suggestion
If you encounter an event resembling the following, it indicates that the Storage Network might have run out its available IPs:
```

Author: Updated per suggestion, thanks.

> If it appears similar event as following, the Storage Network should run out of IPs.

Reviewer: ```suggestion
If the output is similar to the following one, the Storage Network has run out of IPs.
```

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308530799, thanks.

```bash
Events:
  Type     Reason                  Age                    From     Message
  ----     ------                  ----                   ----     -------
  ....

  Warning  FailedCreatePodSandBox  2m58s                  kubelet  Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for
 sandbox "04e9bc160c4f1da612e2bb52dadc86702817ac557e641a3b07b7c4a340c9fc48": plugin type="multus" name="multus-cni-network" failed (add): [longhorn-system/ba
cking-image-ds-default-image-lxq7r/7d6995ee-60a6-4f67-b9ea-246a73a4df54:storagenetwork-sdfg8]: error adding container to network "storagenetwork-sdfg8": erro
r at storage engine: Could not allocate IP in range: ip: 172.16.0.1 / - 172.16.0.6 / range: net.IPNet{IP:net.IP{0xac, 0x10, 0x0, 0x0}, Mask:net.IPMask{0xff,
0xff, 0xff, 0xf8}}

  ....
```

> Please reconfigure Storage Network with sufficient IP range.

Reviewer: ```suggestion
Please reconfigure the Storage Network with a sufficient IP range.
```

Author: Updated per suggestion, thanks.

:::note

> The same error could be encountered during image uploading/downloading, if Storage Network is out of IPs. Please delete related images, and reconfigure Storage Network with sufficient IP range.

Reviewer: ```suggestion
If the Storage Network has run out of IPs, you might encounter the same error during image uploads/downloads. In this case, delete the related images and reconfigure the Storage Network with a sufficient IP range.
```

Author: Updated per suggestion, thanks.

:::

#### Step 3

> Check if Annotations `k8s.v1.cni.cncf.io/network-status` has an interface named `lhnet1` and the IP address is in the IP range.

Reviewer 2: ```suggestion
Check the `k8s.v1.cni.cncf.io/network-status` annotations and ensure that an interface named `lhnet1` exists, with an IP address within the designated IP range.
```

Author: Updated per suggestion, thanks.

> Check if Annotations `k8s.v1.cni.cncf.io/network-status` has an interface named `lhnet1` and the IP address is in the IP range.

Reviewer: ```suggestion
Check if the annotation `k8s.v1.cni.cncf.io/network-status` has an interface named `lhnet1` and the IP address is in the IP range.
```

Author: duplicated as https://github.com/harvester/docs/pull/406#discussion_r1308538931, thanks.

#### Step 2

Verify the readiness of all Longhorn `instance-manager-e`, `instance-manager-r`, and `backing-image-manager` pods, and confirm that their networks are correctly configured.

Execute the following command to inspect a pod's details:

```bash
kubectl -n longhorn-system describe pod <pod-name>
```

> If you encounter an event resembling the following, it indicates that the Storage Network might have run out its available IPs:

Reviewer: ```suggestion
If you encounter an event resembling the following one, the Storage Network might have run out of its available IPs:
```

```bash
Events:
  Type     Reason                  Age                    From     Message
  ----     ------                  ----                   ----     -------
  ....

  Warning  FailedCreatePodSandBox  2m58s                  kubelet  Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for
 sandbox "04e9bc160c4f1da612e2bb52dadc86702817ac557e641a3b07b7c4a340c9fc48": plugin type="multus" name="multus-cni-network" failed (add): [longhorn-system/ba
cking-image-ds-default-image-lxq7r/7d6995ee-60a6-4f67-b9ea-246a73a4df54:storagenetwork-sdfg8]: error adding container to network "storagenetwork-sdfg8": erro
r at storage engine: Could not allocate IP in range: ip: 172.16.0.1 / - 172.16.0.6 / range: net.IPNet{IP:net.IP{0xac, 0x10, 0x0, 0x0}, Mask:net.IPMask{0xff,
0xff, 0xff, 0xf8}}

  ....
```

Please reconfigure the Storage Network with a sufficient IP range.

:::note

> The same error could be encountered during image uploading/downloading, if Storage Network is out of IPs. Please delete related images, and reconfigure Storage Network with sufficient IP range.

Reviewer: ```suggestion
If the Storage Network has run out of IPs, you might encounter the same error when you upload/download images. Please delete the related images and reconfigure the Storage Network with a sufficient IP range.
```
